src/analyserexternalvariable.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/analyserexternalvariable.h"
+
+#include "libcellml/component.h"
+
+#include "analyserexternalvariable_p.h"
+#include "utilities.h"
+
+namespace libcellml {
+
+AnalyserExternalVariable::AnalyserExternalVariableImpl::AnalyserExternalVariableImpl(const VariablePtr &variable)
+    : mVariable(variable)
+{
+}
+
+std::vector<VariableWeakPtr>::iterator AnalyserExternalVariable::AnalyserExternalVariableImpl::findDependency(const ModelPtr &model,
+                                                                                                              const std::string &componentName,
+                                                                                                              const std::string &variableName)
+{
+    return std::find_if(mDependencies.begin(), mDependencies.end(), [=](const VariableWeakPtr &v) {
+        auto variable = v.lock();
+
+        return (variable != nullptr)
+               && (owningModel(variable) == model)
+               && (owningComponent(variable)->name() == componentName)
+               && (variable->name() == variableName);
+    });
+}
+
+std::vector<VariableWeakPtr>::iterator AnalyserExternalVariable::AnalyserExternalVariableImpl::findDependency(const VariablePtr &variable)
+{
+    return std::find_if(mDependencies.begin(), mDependencies.end(), [=](const VariableWeakPtr &v) {
+        return v.lock() == variable;
+    });
+}
+
+AnalyserExternalVariable::AnalyserExternalVariable(const VariablePtr &variable)
+    : mPimpl(new AnalyserExternalVariableImpl(variable))
+{
```

Comment:
Are we talking about the `AnalyserExternalVariable` class? If so, where is it used to create an instance of `AnalyserExternalVariable` without specifying an argument?

Reply:
That's my point.  That even though it might not be used, we have a different form of construction than elsewhere?  eg: we have to create a variable and then add it to the component, as opposed to creating a variable with a component as an argument?  It's just not the same as we've done it elsewhere?

Reply:
It's quite different. An `AnalyserExternalVariable` needs to be linked to a `Variable`, but a `Variable` doesn't need to know (and doesn't care) about the `AnalyserExternalVariable` that is linked to it. However, an `Analyser` needs to know about `AnalyserExternalVariable`s. This is the reason we have `Analyser::addExternalVariable()` in the same way that we have `Component::addVariable()`.

Reply:
Considering this conversation resolved, but feel free to unresolve it, if needed.